image/gimp_brush.ksy

```diff
@@ -0,0 +1,32 @@
+meta:
+  id: gimp_brush
+  title: GIMP (GNU Image Manipulation Program) brush version 2 file
+  file-extension: gbr
+  license: CC0-1.0
+  endian: be
+doc-ref: https://gitlab.gnome.org/GNOME/gimp/-/blob/master/devel-docs/gbr.txt
+seq:
+  - id: header_size
+    type: u4be
+  - id: version
+    type: u4be
+  - id: width
+    type: u4be
+  - id: height
+    type: u4be
+  - id: color_depth
+    type: u4be
+  - id: magic
+    contents: GIMP
+  - id: spacing
+    type: u4be
+  - id: brush_name
+    type: strz
+    size: header_size-1 - 28
+    encoding: UTF-8
+instances:
+  body_size:
+    value: width * height * color_depth
```

Comment:
Please consider structuring it, since now it is not clear, what is the order of the pixels and components within pixels.

Reply:
I am not sure I entirely understand this one (this is my first .ksy, so please bear with me). It is basically a row based format. Depending on the color depth there might be a single byte (grayscale) or four bytes (RGBA). How would you structure this?

Reply:
```yaml
seq:
...
  - id: body
    type: row
    repeat: expr
    repeat-expr: header.height

types:
  row:
    seq:
      - id: pixels
        type:
          switch-on: _root.header.color_depth
          cases:
            color_depth::grayscale: pixel_gray
            color_depth::rgba: pixel_rgba
        repeat: expr
        repeat-expr: _root.header.width
    types:
      pixel_gray:
        seq:
          - id: gray
            type: 1
        instances:
          red:
            value: gray
          green:
            value: gray
          blue:
            value: gray
          alpha:
            value: 1

      pixel_rgba:
        seq:
          - id: red
            type: u1
          - id: green
            type: u1
          - id: blue
            type: u1
          - id: alpha
            type: u1
```

Reply:
Thank you @KOLANICH for the pixel structures! They work fine except the `pixel_gray` type, which should be like this instead:

```ksy
      pixel_gray:
        seq:
          - id: gray
            type: u1
        instances:
          red:
            value: 0
          green:
            value: 0
          blue:
            value: 0
          alpha:
            value: gray
```

I found this experimentally: after you edit the pixel value (actually the last byte) of the brush [`Basic/pixel.gbr`](https://gitlab.gnome.org/GNOME/gimp/-/blob/0ef5d67/data/brushes/Basic/pixel.gbr) in a hex editor (e.g. to `0x55` or `85` in decimal), load it into GIMP and try to draw something with it (but I recommend setting _Size_ to some large value like 50, because with the default size 1 the edges of the brush stroke will be smoothed and you might not be able to pick up the right color).

Then if you press the <kbd>O</kbd> (as in **O**scar, not zero: <kbd>0</kbd>) key to select the Color Picker and pick the color while holding <kbd>Shift</kbd>, the info window that pops up should show this:

<table>
<thead>
<tr>
<th colspan=3 align=center>Color Picker</th>
</tr>
<tr>
<th colspan=2>Pixel</th>
<th>RGB (%)</th>
</tr>
</thead>
<tbody>
<tr>
<td align=left>R:</td>
<td align=right>0</td>
<td rowspan=4 align=center>...</td>
</tr>
<tr>
<td align=left>G:</td>
<td align=right>0</td>
</tr>
<tr>
<td align=left>B:</td>
<td align=right>0</td>
</tr>
<tr>
<td align=left>A:</td>
<td align=right>85</td>
</tr>
</tbody>
</table>

---

However, although it kind of makes sense to _describe_ all structures in the KS language that can be described in it (at least for documentation/educational purposes), the question is if they are usable in the real world scenarios (whether the performance penalty and memory usage will be acceptable even for larger files).

I haven't measured anything, but I suspect that creating a class instance for each pixel will slightly (I estimate 3-4 times) increase the memory consumption and the overhead with creating them will probably slow down the parsing a bit (in comparison with creating a fixed-size byte array in advance and just gradually filling it with RGBA values as you parse the bitmap, just as you would've done when parsing the file with a hand-written code). Naturally, it would be best to have https://github.com/kaitai-io/kaitai_struct/issues/188, which would be ultimately exactly just filling the fixed-size buffer, but I wonder if the parsing of bitmap pixels is feasible using purely (or _mostly_) KS in the current state.

Reply:
>However, although it kind of makes sense to describe all structures in the KS language that can be described in it (at least for documentation/educational purposes), the question is if they are usable in the real world scenarios (whether the performance penalty and memory usage will be acceptable even for larger files).

Absolutely valid concern. The solution may be to mark some types and fields as "hardware-implemented", so the compiler won't usually generate code for them, but to keep them as a blob.

And an another mitigation is to redesign runtimes to have structs with less overhead for simple cases (i.e. when `_io` is not needed) and make the compiler to use them and generate the code allocating them properly.

About kaitai-io/kaitai_struct#188 : I don't really beleive in it for images, because there are quite some standard formats for pixels (with different encodings, not necessarily RGB, HSV or YCrCb), and there may be a lot of non-standard, especially in old games. It is impossible and makes no sense to support them all in KS. (Though I guess it may be possible to create a machine-readable spec DSL for pixel formats and let people encode them themeself, but I guess it is out of scope of KS project)